File: modulo2/shopapp/app/src/main/java/com/shopapp/presentation/ui/auth/LoginScreen.kt
// presentation/ui/auth/LoginScreen.kt
package com.shopapp.presentation.ui.auth

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.hilt.navigation.compose.hiltViewModel
import com.shopapp.presentation.components.*
import com.shopapp.presentation.viewmodel.AuthViewModel
import com.shopapp.theme.*

@Composable
fun LoginScreen(
    onLoginSuccess:  (isStaff: Boolean) -> Unit,
    onNavigateToRegister: () -> Unit,
    viewModel: AuthViewModel = hiltViewModel(),
) {
    val uiState by viewModel.uiState.collectAsState()

    var username by remember { mutableStateOf("") }
    var password by remember { mutableStateOf("") }

    // Navegar cuando el login es exitoso
    LaunchedEffect(uiState) {
        if (uiState is AuthUiState.Success) {
            val user = (uiState as AuthUiState.Success).user
            onLoginSuccess(user.isStaff)
        }
    }

    val isLoading = uiState is AuthUiState.Loading
    val errorMsg  = (uiState as? AuthUiState.Error)?.message

    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(Background),
    ) {
        Column(
            modifier = Modifier
                .fillMaxSize()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 24.dp)
                .padding(top = 80.dp, bottom = 40.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
        ) {
            // Logo
            Text(
                text       = "ShopApp",
                fontSize   = 36.sp,
                fontWeight = FontWeight.Bold,
                color      = Accent,
            )
            Spacer(Modifier.height(8.dp))
            Text(
                text  = "Inicia sesión en tu cuenta",
                color = TextSecondary,
                style = MaterialTheme.typography.bodyMedium,
            )
            Spacer(Modifier.height(40.dp))

            // Card del formulario
            Surface(
                shape            = MaterialTheme.shapes.large,
                color            = Surface,
                tonalElevation   = 0.dp,
                modifier         = Modifier.fillMaxWidth(),
            ) {
                Column(modifier = Modifier.padding(24.dp)) {

                    // Error general
                    if (errorMsg != null) {
                        Surface(
                            color  = Error.copy(alpha = 0.1f),
                            shape  = MaterialTheme.shapes.small,
                            modifier = Modifier.fillMaxWidth(),
                        ) {
                            Text(
                                text     = errorMsg,
                                color    = Error,
                                style    = MaterialTheme.typography.bodySmall,
                                modifier = Modifier.padding(12.dp),
                            )
                        }
                        Spacer(Modifier.height(16.dp))
                    }

                    // Campo usuario
                    ShopTextField(
                        value         = username,
                        onValueChange = { username = it; viewModel.clearError() },
                        label         = "Usuario",
                        placeholder   = "tu_usuario",
                        enabled       = !isLoading,
                        imeAction     = ImeAction.Next,
                    )
                    Spacer(Modifier.height(16.dp))

                    // Campo contraseña
                    ShopTextField(
                        value         = password,
                        onValueChange = { password = it; viewModel.clearError() },
                        label         = "Contraseña",
                        placeholder   = "••••••••",
                        isPassword    = true,
                        enabled       = !isLoading,
                        keyboardType  = KeyboardType.Password,
                        imeAction     = ImeAction.Done,
                    )
                    Spacer(Modifier.height(24.dp))

                    // Botón
                    ShopButton(
                        text      = "Iniciar sesión",
                        onClick   = { viewModel.login(username, password) },
                        isLoading = isLoading,
                        enabled   = username.isNotBlank() && password.isNotBlank(),
                    )
                }
            }

            Spacer(Modifier.height(24.dp))

            // Link a registro
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text(
                    text  = "¿No tienes cuenta? ",
                    color = TextSecondary,
                    style = MaterialTheme.typography.bodyMedium,
                )
                TextButton(onClick = onNavigateToRegister) {
                    Text(
                        text  = "Regístrate",
                        color = Accent,
                        style = MaterialTheme.typography.bodyMedium,
                        fontWeight = FontWeight.SemiBold,
                    )
                }
            }
        }
    }
}

File: modulo4/flutter_shop_app/lib/presentation/screens/auth/login_screen.dart
// lib/presentation/screens/auth/login_screen.dart

import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:go_router/go_router.dart';
import '../../../theme/app_colors.dart';
import '../../../core/utils/validators.dart';
import '../../../domain/model/auth_state.dart';
import '../../providers/auth_provider.dart';
import '../../widgets/auth_button.dart';
import '../../widgets/auth_text_field.dart';

class LoginScreen extends ConsumerStatefulWidget {
  const LoginScreen({super.key});

  @override
  ConsumerState<LoginScreen> createState() => _LoginScreenState();
}

class _LoginScreenState extends ConsumerState<LoginScreen> {
  final _formKey    = GlobalKey<FormState>();
  final _userCtrl   = TextEditingController();
  final _passCtrl   = TextEditingController();
  bool  _submitted  = false;

  @override
  void dispose() {
    _userCtrl.dispose();
    _passCtrl.dispose();
    super.dispose();
  }

  Future<void> _submit() async {
    setState(() => _submitted = true);
    if (!_formKey.currentState!.validate()) return;
    await ref.read(authProvider.notifier).login(
      _userCtrl.text,
      _passCtrl.text,
    );
  }

  @override
  Widget build(BuildContext context) {
    final authState = ref.watch(authProvider);
    final isLoading = authState.isChecking;
    final error     = authState.error;
    final tt        = Theme.of(context).textTheme;

    // Escuchar cambios de estado para navegar
    ref.listen<AuthState>(authProvider, (_, next) {
      if (next.isAuthenticated) {
        final dest = next.isStaff ? '/admin' : '/';
        context.go(dest);
      }
    });

    return Scaffold(
      body: SafeArea(
        child: SingleChildScrollView(
          padding: const EdgeInsets.symmetric(horizontal: 24),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.center,
            children: [
              const SizedBox(height: 80),

              // Logo
              Text(
                'Flutter Shop App',
                style: tt.displayMedium?.copyWith(color: AppColors.accent),
              ),
              const SizedBox(height: 8),
              Text('Inicia sesión en tu cuenta', style: tt.bodyMedium),
              const SizedBox(height: 48),

              // Card del formulario
              Container(
                padding:    const EdgeInsets.all(24),
                decoration: BoxDecoration(
                  color:        AppColors.surface,
                  borderRadius: BorderRadius.circular(20),
                  border:       Border.all(color: AppColors.border),
                ),
                child: Form(
                  key: _formKey,
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [

                      // Error general del servidor
                      if (error != null) ...[
                        Container(
                          width:   double.infinity,
                          padding: const EdgeInsets.all(12),
                          decoration: BoxDecoration(
                            color:        AppColors.error.withValues(alpha: 0.1),
                            borderRadius: BorderRadius.circular(10),
                            border:       Border.all(color: AppColors.error.withValues(alpha: 0.3)),
                          ),
                          child: Text(
                            error,
                            style: const TextStyle(color: AppColors.error, fontSize: 13),
                          ),
                        ),
                        const SizedBox(height: 16),
                      ],

                      // Campo usuario
                      AuthTextField(
                        label:      'Usuario',
                        hint:       'tu_usuario',
                        controller: _userCtrl,
                        enabled:    !isLoading,
                        validator:  _submitted ? validateUsername : null,
                        onChanged:  (_) => ref.read(authProvider.notifier).clearError(),
                      ),
                      const SizedBox(height: 16),

                      // Campo contraseña
                      AuthTextField(
                        label:           'Contraseña',
                        hint:            '••••••••',
                        controller:      _passCtrl,
                        isPassword:      true,
                        enabled:         !isLoading,
                        keyboardType:    TextInputType.visiblePassword,
                        textInputAction: TextInputAction.done,
                        validator:       _submitted
                            ? (v) => (v == null || v.isEmpty) ? 'Campo obligatorio' : null
                            : null,
                        onChanged:       (_) => ref.read(authProvider.notifier).clearError(),
                      ),
                      const SizedBox(height: 24),

                      // Botón
                      AuthButton(
                        label:     'Iniciar sesión',
                        onPressed: _submit,
                        isLoading: isLoading,
                      ),
                    ],
                  ),
                ),
              ),

              const SizedBox(height: 24),

              // Link al registro
              Row(
                mainAxisAlignment: MainAxisAlignment.center,
                children: [
                  Text('¿No tienes cuenta? ', style: tt.bodyMedium),
                  TextButton(
                    onPressed: () => context.push('/register'),
                    child: const Text('Regístrate'),
                  ),
                ],
              ),
            ],
          ),
        ),
      ),
    );
  }
}